### PR TITLE
Why use json when you can use python?

### PR DESCRIPTION
**Pull request by gwatt**
Instead of storing your config as a json object in a file, just use
python! Json is a pain to edit by hand, all those stupid braces and
random quotes. Python is much nicer. A little work has to be done in the
loading file to ensure all requisite values have reasonable defaults,
but that's not too hard.

### CODE REVIEW

File: pucauto.py [hunk 1]
```diff
@@ -10,10 +10,21 @@
 from bs4 import BeautifulSoup
 from lib import logger
 
-
-with open("config.json") as config:
-    CONFIG = json.load(config)
-
+try:
```
Comment (tomreece) [inline pucauto.py]: I'd prefer to remove this try/catch and just let it blow up if they don't have a config.py. Less code. They'll still get an error mentioning config.py.

File: pucauto.py [hunk 2]
```diff
@@ -10,10 +10,21 @@
...
+    import config
+except:
+    print 'No config.py found!'
+    exit()
+
+config_vars = dir(config)
+for var in ['username', 'password', 'min_value']:
```
Comment (tomreece) [inline pucauto.py]: I'd also prefer to remove this and just let the script blow up if they don't have a valid config.py. Less code. Not my fault they didn't match the example. :stuck_out_tongue:

File: pucauto.py [hunk 3]
```diff
@@ -10,10 +10,21 @@
...
+    if var not in config_vars:
+        print "Required configuration '%s' not found!" % (var,)
+        exit()
+if 'find_add_ons' not in config_vars:
```
Comment (tomreece) [inline pucauto.py]: I guess this is one thing I dislike about the .py approach vs the .json approach. I can't just call CONFIG.get(...) and if it returns `None` then make decisions. Now I have to provide defaults for everything at the top of the script. Hmm.

File: pucauto.py [hunk 4]
```diff
@@ -91,8 +102,7 @@ def check_runtime():
     RAM cloud server.
     """
 
-    hours_to_run = CONFIG.get("hours_to_run")
-    if hours_to_run:
+    if config.hours_to_run:
         return (datetime.now() - START_TIME).total_seconds() / 60 / 60 < hours_to_run
```
Comment (tomreece) [inline pucauto.py]: Need `config.hours_to_run` on this line too.

File: pucauto.py [hunk 5]
```diff
@@ -101,8 +111,7 @@ def check_runtime():
 def should_check_add_ons():
     """Return True if we should check for add on trades."""
 
-    minutes_between_add_ons_check = CONFIG.get("minutes_between_add_ons_check")
-    if minutes_between_add_ons_check:
+    if cofnig.minutes_between_add_ons_check:
```
Comment (tomreece) [inline pucauto.py]: Typo in `cofnig`

File: pucauto.py [hunk 6]
```diff
@@ -101,8 +111,7 @@ def check_runtime():
...
         return (datetime.now() - LAST_ADD_ON_CHECK).total_seconds() / 60 >= minutes_between_add_ons_check
```
Comment (tomreece) [inline pucauto.py]: Need `config.minutes_between_add_ons_check` here too.

**Comment by tomreece** [comment]
You'd also need to update the `README.md` documentation to update references to `config.json` or `config.example.json`.

Aside from that, I added various small comments. I think I prefer the .json approach better for some reason. Mainly so I can add more config options in the future and rely on `.get()` returning `None` instead of having to remember to provide defaults.

Is there any great reason you give me to switch to the .py approach, other than the fact that .json has some quotes and commas?

**Comment by gwatt** [comment]
Not that it really matters for this, but if you make your config python source code, you can have logic in the config.

**Comment by eengstrom** [comment]
> Is there any great reason you give me to switch to the .py approach, other than the fact that .json has some quotes and commas?

Although I'm ambivalent about this change in this context, my biggest beef with JSON is the lack of comments - I mean, who doesn't need a little explanation at times?  JSON as an interchange format?  Fine.  JSON as a config file?  Ugh.

**Comment by tomreece** [comment]
@eengstrom good point

**Comment by eengstrom** [comment]
Ok, so how about [ConfigParser](https://docs.python.org/2/library/configparser.html) ([example here](https://wiki.python.org/moin/ConfigParserExamples)).  That way you get comments and interpolation of values, but not full conditional/code execution.  You also get the simplicity of JSON w/out the extra code.  It's also a base module, so no extra dependencies.

Note - one potential issue with leading spaces, though I don't think that's true; see last comment in:
http://stackoverflow.com/questions/1720057/git-config-style-configuration-system

**Comment by gwatt** [comment]
I've filled out the config example to contain all necessary fields and removed error handling from pucauto.py.

**Comment by tomreece** [comment]
@gwatt There are still a few references to variables that aren't assigned. https://github.com/tomreece/pucauto/pull/52/files

**Comment by tomreece** [comment]
@eengstrom @gwatt what if I just used something like this for json comments? https://commentjson.readthedocs.org/en/latest/ It lets you write:

```
{
    "name": "Vaidik Kapoor", # Person's name
    "location": "Delhi, India", // Person's location

    # Section contains info about
    // person's appearance
    "appearance": {
        "hair_color": "black",
        "eyes_color": "black",
        "height": "6"
    }
}
```

**Comment by gwatt** [comment]
I don't have a very strong opinion. It's your project and you should do what works best for you. This PR came out of me idly asking you what you thought about using python as a config language over json. Do with it what you will.

**Comment by tomreece** [comment]
The biggest problem users are having with .json as the config file, is that their editors (namely TextEdit on Mac) are changing the quotes to smart quotes, which would still happen if I use a .py file for config.

**Pull request closed without merging**